bindings/python/src/lib.rs

```diff
@@ -211,6 +211,144 @@ struct safe_open {
+                    exceptions::PyException::new_err(format!("Couldn't load cuda {e:?}",))
+                })?;
+                return Ok(lib);
+            }
+        }
+    }
+    Err(exceptions::PyException::new_err("Couldn't find cuda"))
+}
+
+fn create_cuda_unsafe_tensor(
+    module: &PyModule,
+    cudart: &libloading::Library,
+    info: &TensorInfo,
+    device: &Device,
+    data: &[u8],
+) -> PyResult<PyObject> {
+    let tensor = create_empty_tensor_pt(module, &info.shape, info.dtype, device)?;
+
+    let data_ptr_fn = tensor.getattr("data_ptr")?;
+    let data_ptr: usize = data_ptr_fn.call0()?.extract()?;
+
+    // SAFETY: This is unsafe for the same reasons as when we load the library
+    let out = unsafe {
+        let cuda_memcpy: libloading::Symbol<
+            unsafe extern "C" fn(
+                device_ptr: u64,
+                src_ptr: *const std::ffi::c_void,
+                src_len: usize,
+            ) -> u32,
+        > = cudart.get(b"cudaMemcpy").map_err(|e| {
+            exceptions::PyException::new_err(format!("Couldn't find cudaMemcpy {e:?}",))
+        })?;
+        cuda_memcpy(
+            data_ptr as u64,
+            data.as_ptr() as *const std::ffi::c_void,
+            data.len(),
+        )
+    };
+    // SAFETY: Here we have a correct library, we successfully called memcpy,
+    // but somehow the call failed. This is really worrying since Pytorch is
+    // responsible for allocating the memory.
+    if out != 0 {
+        panic!(
+            "We tried to set your tensor fast, but there was a cuda error, This could
+                have corrupted your GPU ram, aborting to prevent further errors"
+        )
+    }
+    let tensor: PyObject = tensor.into_py(module.py());
+    Ok(tensor)
+}
+
+fn create_cuda_unsafe_tensor_from_slice(
+    module: &PyModule,
+    cudart: &libloading::Library,
+    shape: &[usize],
+    dtype: Dtype,
+    device: &Device,
+    iterator: safetensors::slice::SliceIterator,
+) -> PyResult<PyObject> {
+    let tensor = create_empty_tensor_pt(module, shape, dtype, device)?;
+
+    let data_ptr_fn = tensor.getattr("data_ptr")?;
+    let data_ptr: usize = data_ptr_fn.call0()?.extract()?;
+    let mut offset = 0;
+    // SAFETY: This is unsafe for the same reasons as when we load the library
+    unsafe {
+        let cuda_memcpy: libloading::Symbol<
+            unsafe extern "C" fn(
+                device_ptr: u64,
+                src_ptr: *const std::ffi::c_void,
+                src_len: usize,
+            ) -> u32,
+        > = cudart.get(b"cudaMemcpy").map_err(|e| {
+            exceptions::PyException::new_err(format!("Couldn't find cudaMemcpy {e:?}",))
+        })?;
+        for slice in iterator {
+            let len = slice.len();
+            let out = cuda_memcpy(
+                (data_ptr + offset) as u64,
+                slice.as_ptr() as *const std::ffi::c_void,
+                len,
+            );
```

Comment:
Personally I like to reduce the scope of the `unsafe`, so I'd just put it around here and the `cudart.get` call.

```suggestion
            // Safety: pointer and len are guaranteed valid as they originate from the same slice
            let out = unsafe { cuda_memcpy(
                (data_ptr + offset) as u64,
                slice.as_ptr() as *const std::ffi::c_void,
                len,
            ) };
```

Reply:
I agree in general. Here I tried to encapsulate both fetching the symbol and the various calls, as basically they all influence each other.